src/python_be.cc

```diff
@@ -981,83 +388,24 @@ ModelInstanceState::SaveRequestsToSharedMemory(
 }
 
 TRITONSERVER_Error*
-ModelInstanceState::SetupStubProcess()
+ModelInstanceState::LaunchModelInstanceStubProcess()
 {
-  std::string kind = TRITONSERVER_InstanceGroupKindString(kind_);
   ModelState* model_state = reinterpret_cast<ModelState*>(Model());
-
-  // Increase the stub process count to avoid shared memory region name
-  // collision
-  model_state->StateForBackend()->number_of_instance_inits++;
-  shm_region_name_ =
-      model_state->StateForBackend()->shared_memory_region_prefix +
-      std::to_string(model_state->StateForBackend()->number_of_instance_inits);
-
-  uint64_t model_version = model_state->Version();
-  const char* model_path = model_state->RepositoryPath().c_str();
-
-  std::stringstream ss;
-  std::string artifact_name;
-  RETURN_IF_ERROR(model_state->ModelConfig().MemberAsString(
-      "default_model_filename", &artifact_name));
-  ss << model_path << "/" << model_version << "/";
-
-  if (artifact_name.size() > 0) {
-    ss << artifact_name;
-  } else {
-    // Default artifact name.
-    ss << "model.py";
-  }
-
-  model_path_ = ss.str();
-  struct stat buffer;
-
-  // Check if model.py exists
-  if (stat(model_path_.c_str(), &buffer) != 0) {
-    return TRITONSERVER_ErrorNew(
-        TRITONSERVER_ERROR_INTERNAL,
-        ("model.py does not exist in the model repository path: " + model_path_)
-            .c_str());
-  }
-
-  // Path to the extracted Python env
-  std::string python_execution_env = "";
-  if (model_state->PythonExecutionEnv() != "") {
-    try {
-      python_execution_env =
-          model_state->StateForBackend()->env_manager->ExtractIfNotExtracted(
-              model_state->PythonExecutionEnv());
-    }
-    catch (PythonBackendException& pb_exception) {
-      return TRITONSERVER_ErrorNew(
-          TRITONSERVER_ERROR_INTERNAL, pb_exception.what());
-    }
-
-    path_to_activate_ = python_execution_env + "/bin/activate";
-    path_to_libpython_ = python_execution_env + "/lib";
-    if (python_execution_env.length() > 0 && !FileExists(path_to_activate_)) {
-      return TRITONSERVER_ErrorNew(
-          TRITONSERVER_ERROR_INTERNAL,
-          (std::string("Path ") + path_to_activate_ +
-           " does not exist. The Python environment should contain an "
-           "'activate' script.")
-              .c_str());
-    }
-  }
-
-  parent_pid_ = getpid();
-  RETURN_IF_ERROR(StartStubProcess());
+  Stub() = std::make_unique<StubLauncher>(
+      "MODEL_INSTANCE_STUB", Name(), DeviceId(),
+      TRITONSERVER_InstanceGroupKindString(Kind()));
+  RETURN_IF_ERROR(Stub()->Initialize(model_state));
+  RETURN_IF_ERROR(Stub()->Launch());
 
   if (model_state->IsDecoupled()) {
     decoupled_thread_ = true;
-    decoupled_monitor_ =
+    *Stub()->DecoupledMonitor() =
```

Comment:
I think the decoupled monitor doesn't need to be part of the stub launcher and can remain in model instance state.